aya/src/symbols.rs

```diff
@@ -0,0 +1,688 @@
+            (Some(object_path), Some(object_address)) => (object_path, object_address),
+            _ => return,
+        };
+
+        let mut guard = self.state.borrow_mut();
+        symbol.function_name = match guard.object_resolvers.get_mut(object_path) {
+            Some(Some(p)) => p.resolve(object_address),
+            Some(None) => return,
+            None => match Addr2LineObjectContext::from_object_path(object_path) {
+                Err(_) => {
+                    // Cache the error, in order to avoid retrying loading the file
+                    guard.object_resolvers.insert(object_path.to_owned(), None);
+                    return;
+                }
+                Ok(addr2line) => {
+                    let function_name = addr2line.resolve(object_address);
+                    guard
+                        .object_resolvers
+                        .insert(object_path.to_owned(), Some(addr2line));
+                    function_name
+                }
+            },
+        };
+    }
+}
+
+/// A [`SymbolResolver`] which uses the [`Addr2LineResolver`] for resolving
+/// userspace functions, and the [`KernelSymbolResolver`] for resolving kernel functions.
+struct CombinedResolver {
+    kernel: KernelSymbolResolver,
+    user: Addr2LineResolver,
+}
+
+impl CombinedResolver {
+    pub fn new() -> Result<Self, CreateResolverError> {
+        let kernel = KernelSymbolResolver::new()?;
+        let user = Addr2LineResolver::builder().build()?;
+        Ok(Self { kernel, user })
+    }
+}
+
+impl SymbolResolver for CombinedResolver {
+    fn resolve(&self, symbol: &mut SymbolInfo) {
+        match symbol.is_user() {
+            true => self.user.resolve(symbol),
+            false => self.kernel.resolve(symbol),
+        }
+    }
+}
+
+/// A default resolver which can resolve kernel and userspace symbols,
+/// and caches last observed symbol information in memory.
+///
+/// # Examples
+///
+/// ```no_run
+/// # use aya::symbols::{DefaultResolver, SymbolResolver, SymbolInfo};
+/// let resolver = DefaultResolver::new().unwrap();
+///
+/// // Resolve a kernel-space symbol based on an address obtained from a stack trace
+/// let mut kernel_symbol = SymbolInfo::unresolved_kernel(0x1234_5678);
+/// resolver.resolve(&mut kernel_symbol);
+/// println!("{:?}", kernel_symbol.function_name());
+///
+/// // Resolve a user-space symbol based on an a process-id and address obtained from a stack trace
+/// let mut user_symbol = SymbolInfo::unresolved_user(7654, 0x1234_5678);
+/// resolver.resolve(&mut user_symbol);
+/// println!("{:?}", user_symbol.function_name());
+/// ```
+pub struct DefaultResolver {
+    inner: CachingResolver<CombinedResolver>,
+}
+
+impl DefaultResolver {
+    /// Create a new DefaultResolver
+    pub fn new() -> Result<Self, CreateResolverError> {
+        let caching_resolver = CachingResolver::with_capacity(8192, CombinedResolver::new()?);
```

Comment:
I have no strong opinion about any of the LRU capacities in this change. I just thought it would be a good idea to limit the amount of metadata that the process loads - if this is used for long-running profiling.